Deprecate metadata as additional field kwargs, support explicit `metadata=...`

sirosen opened the pull request:
I'm opening this to kickstart discussion (and volunteer to do any necessary work). I haven't rewritten any tests or added new tests or updated docs yet.

Add `metadata=...` as an explicit keyword argument, which dict-ifies a mapping, and treat any additional kwargs as "added metadata" which triggers a deprecation warning.

This is backwards compatible *except* in the case where someone is already using `metadata` as the name of a metadata field. Which is probably never intentional if it's done at all.

resolves #1350

---

Questions/TODOs on this:

- [x] Is it okay to add `metadata=...` to kwargs in 3.x or does this whole change have to wait for 4.x ? (I don't think so, maybe others disagree.)
- [x] Is doing the shallow-copy with `dict(metadata)` the right thing? Should we maybe do nothing with the input?
- [x] Rewrite tests to use explicit `metadata` now
- [x] Update docs within `marshmallow`
- [x] Should I prepare PRs on apispec and other `marshmallow-code` projects to update `metadata` style in docs when this merges?

sirosen committed:
Deprecate metadata as additional field kwargs

Add `metadata=...` as an explicit keyword argument, which dict-ifies a
mapping, and treat any additional kwargs as "added metadata" which
triggers a deprecation warning.

This is backwards compatible *except* in the case where someone is
already using `metadata` as the name of a metadata field. Which is
probably never intentional if it's done at all.

resolves #1350

lafrech commented:
>  Is doing the shallow-copy with dict(metadata) the right thing? Should we maybe do nothing with the input?

Yes. As long as we update with the extra kwargs. We don't want to modify the input dict.

sloria commented on src/marshmallow/fields.py:
```diff
@@ -187,7 +188,19 @@ def __init__(
         self.missing = missing
-        self.metadata = metadata
+
+        # intentionally shallow-copy so that more data can be added without changing
+        # the original mapping object, but objects can be referenced from field
+        # metadata
+        self.metadata = dict(metadata) if metadata is not None else {}
+        if additional_metadata:
+            self.metadata.update(additional_metadata)
```
Maybe

```suggestion
        metadata = metadata or {}
        self.metadata = {**metadata, **additional_metadata}
        if additional_metadata:
```

sirosen replied:
I like this better! Taking it.

Also, I didn't know that `**` could be used on an arbitrary Mapping until today. So thanks for teaching me something new. 😄

sirosen committed:
Simplify Field.metadata initialization

Rather than explicitly dict-ifying the input and conditionally
running `update()`, combine data with dict expansion.

Co-authored-by: Steven Loria <sloria1@gmail.com>

sirosen committed:
Update tests with explicit field metadata

There is one test of field metadata which changed to use
`metadata={...}` and a single new test which ensures that you can use
either style for setting field metadata, but that if you use the
"arbitrary keyword arguments" pathway, a deprecation warning will be
emitted.

sirosen commented:
I just made the testsuite updates (which are quite modest) for this and looked for any narrative docs which should be updated but didn't find any.

The only question I have left -- since nobody has raised a big concern about doing this in 3.x -- is whether or not I should work on doc updates for apispec and other projects?

sloria commented:
Sorry for the delayed response--yes, I think this is good to go into 3.x.

> I should work on doc updates for apispec and other projects

Good idea; it'd be good to prep for this in webargs and apispec (I think those are the only 2 marshmallow-code projects that use `metadata` as part of their API?).

sirosen commented:
Okay, awesome, I'll get started on apispec.

I don't think there's anything in webargs to change anymore, but I'll double-check.

sirosen commented:
I've opened marshmallow-code/apispec#617 as a counterpart to this, but other than that I don't plan to do more.
So as far as I'm concerned, this is ready. Just let me know if there's more I should do. 🙂
